KAFKA-18923: resource leak in RSM inputStream

FrankYang0529 opened the pull request:
Following are all references about `fetchLogSegment` and `fetchIndex`.

* InputStream fetchLogSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata, int startPosition)
  * RemoteLogManager#lookupTimestamp: https://github.com/apache/kafka/blob/1eea7f0528954ce8dcbcc4357ae2ef28c1d1e5f2/core/src/main/java/kafka/log/remote/RemoteLogManager.java#L653-L655
  * RemoteLogManager#read: https://github.com/apache/kafka/blob/1eea7f0528954ce8dcbcc4357ae2ef28c1d1e5f2/core/src/main/java/kafka/log/remote/RemoteLogManager.java#L1744-L1748
  * _LocalTieredStorageTest#verifyFetchedLogSegment: Fixed by this PR._
* InputStream fetchLogSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata, int startPosition, int endPosition)
  * LocalTieredStorageTest: All usage throws exception before creating input stream.
* InputStream fetchIndex(RemoteLogSegmentMetadata remoteLogSegmentMetadata, IndexType indexType)
  * _TierStateMachine#readLeaderEpochCheckpoint and TierStateMachine#buildProducerSnapshotFile: Fixed by this PR._
  * RemoteIndexCache#createCacheEntry: The input stream is supplied by a function and is closed in RemoteIndexCache#loadIndexFile. https://github.com/apache/kafka/blob/1eea7f0528954ce8dcbcc4357ae2ef28c1d1e5f2/storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java#L355-L357
  * _LocalTieredStorageTest#verifyFetchedOffsetIndex, LocalTieredStorageTest#verifyFetchedTimeIndex, LocalTieredStorageTest#verifyFetchedTransactionIndex, LocalTieredStorageTest#verifyLeaderEpochCheckpoint, and LocalTieredStorageTest#verifyProducerSnapshot: Fixed by this PR. The input stream is closed in LocalTieredStorageTest#verifyFileContents._

FrankYang0529 committed:
KAFKA-18923: resource leak in RSM fetchIndex inputStream

Signed-off-by: PoAn Yang <payang@apache.org>

mumrah committed:
Merge branch 'trunk' into KAFKA-18923

frankvicky reviewed (commented):
It seems that we have a thread leak here.
In my experience, it will lead to `initializationError`
Could you merge the trunk and trigger CI again?

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-18923

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-18923

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-18923

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-18923

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-18923

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-18923

showuon reviewed (commented):
@FrankYang0529 , thanks for the PR. But I can still see some other resource leaks around RSM methods. Ex: [here](https://github.com/apache/kafka/blob/e3809682e2a0a312c6d5594fbc463b6cb0bb2273/storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageTest.java#L486), [here](https://github.com/apache/kafka/blob/e3809682e2a0a312c6d5594fbc463b6cb0bb2273/core/src/main/java/kafka/log/remote/RemoteLogManager.java#L635), .... Could you please help check all the methods usage and make sure all resources are closed correctly?

FrankYang0529 committed:
Merge branch 'trunk' into KAFKA-18923

FrankYang0529 committed:
address comment

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 commented:
Hi @showuon, thanks for the suggestion. Following are all references about `fetchLogSegment` and `fetchIndex`.

* InputStream fetchLogSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata, int startPosition)
  * RemoteLogManager#lookupTimestamp: https://github.com/apache/kafka/blob/1eea7f0528954ce8dcbcc4357ae2ef28c1d1e5f2/core/src/main/java/kafka/log/remote/RemoteLogManager.java#L653-L655
  * RemoteLogManager#read: https://github.com/apache/kafka/blob/1eea7f0528954ce8dcbcc4357ae2ef28c1d1e5f2/core/src/main/java/kafka/log/remote/RemoteLogManager.java#L1744-L1748
  * _LocalTieredStorageTest#verifyFetchedLogSegment: Fixed by this PR._
* InputStream fetchLogSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata, int startPosition, int endPosition)
  * LocalTieredStorageTest: All usage throws exception before creating input stream.
* InputStream fetchIndex(RemoteLogSegmentMetadata remoteLogSegmentMetadata, IndexType indexType)
  * _TierStateMachine#readLeaderEpochCheckpoint and TierStateMachine#buildProducerSnapshotFile: Fixed by this PR._
  * RemoteIndexCache#createCacheEntry: The input stream is supplied by a function and is closed in RemoteIndexCache#loadIndexFile. https://github.com/apache/kafka/blob/1eea7f0528954ce8dcbcc4357ae2ef28c1d1e5f2/storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java#L355-L357
  * _LocalTieredStorageTest#verifyFetchedOffsetIndex, LocalTieredStorageTest#verifyFetchedTimeIndex, LocalTieredStorageTest#verifyFetchedTransactionIndex, LocalTieredStorageTest#verifyLeaderEpochCheckpoint, and LocalTieredStorageTest#verifyProducerSnapshot: Fixed by this PR. The input stream is closed in LocalTieredStorageTest#verifyFileContents._

showuon reviewed (approved):
LGTM! Thanks for the analysis!
